romancal/multiband_catalog/detection_image.py

```diff
@@ -0,0 +1,149 @@
+        The output Gaussian kernel.
+    """
+    size = np.ceil(3 * fwhm).astype(int)
+    size = size + 1 if size % 2 == 0 else size  # make size be odd
+    kernel = make_2dgaussian_kernel(fwhm, size=size)  # sums to 1
+    return kernel.array
+
+
+def make_det_image(library, kernel_fwhm):
+    """
+    Make a detection image from a library of models.
+
+    The detection image is the weighted sum of the input models, where
+    the weights are SED weight divided by the read noise variance.
+
+    Parameters
+    ----------
+    library : `romancal.datamodels.ModelLibrary`
+        The input library of models.
+
+    kernel_fwhm : float
+        The full-width at half-maximum (FWHM) in pixels of the 2D
+        Gaussian kernel used to smooth the detection image.
+
+    Returns
+    -------
+    detection_data : 2D `numpy.ndarray`
+        The detection image data.
+
+    detection_error : 2D `numpy.ndarray`
+        The detection image (standard deviation) error.
+    """
+    if not isinstance(library, ModelLibrary):
+        raise TypeError("library input must be a ModelLibrary object")
+
+    # TODO: extend to different kernel shapes beyond Gaussian?
+    # would require different/additional kernel parameters
+    kernel = make_gaussian_kernel(kernel_fwhm)
+
+    log.info(f"Making detection image with kernel FWHM={kernel_fwhm}")
+
+    with library:
+        detection_data = 0.0
+        detection_var = 0.0
+        wht_sum = 0.0
+        for i, model in enumerate(library):
+            # TODO: SED weights to be defined in the asn file for each
+            # input filter image
+            try:
+                sed_weight = library.asn["products"][0]["members"][i]["sed_weight"]
+            except KeyError:
+                sed_weight = 1.0
+
+            log.info(
+                f"Processing model {model.meta.filename}: "
+                f"filter={model.meta.basic.optical_element}, {sed_weight=}"
+            )
+
+            # Ideally the weights should be the inverse variance of
+            # all sources of noise except the source Poisson noise. The
+            # closest approximation we have is var_rnoise (read noise
+            # variance) -- var_rnoise is also used as the weighting in the
+            # resample step. Note that model.err is the total error, which
+            # includes source Poisson noise.
+            wht = sed_weight / model.var_rnoise  # inverse variance
+            wht_sum += wht
+
+            coverage_mask = np.isnan(model.err)
+            with warnings.catch_warnings():
+                # Suppress warnings about any NaNs in the data
+                warnings.filterwarnings(action="ignore", category=AstropyUserWarning)
+
+                detection_data += convolve_fft(
+                    wht * model.data, kernel, mask=coverage_mask
+                )
+                detection_var += convolve_fft(
+                    wht**2 * model.var_rnoise, kernel, mask=coverage_mask
+                )
```

Comment:
For discussion later, I've been staring at code like this in a different package recently as well, and maybe I am missing something important?  But in my head you have something like:
```
signal(i) = sum_j PSF(i, j) * image(j) * weight(j)
variance(i) = sum_j PSF(i, j)^2 * sigma^2(j) * weight^2(j)
```
this code has the sigma^2 and weight^2 terms but not the PSF^2 term.

Reply:
I saw in your doc file the `kernel**2` term, but I have questions about it.  Since `kernel` is normalized to sum to 1, convolving the variance with `kernel**2` does not conserve the variance.  Convolution is a linear operation, so why would you want to square the kernel?

Reply:
I claim that the detection significance image should be the local significance of the kernel at each location.  This is a linear fit with a profile P.  For a normal linear least squares problem, if the uncertainties were given by a covariance matrix C, the best fit fluxes would be:
x = (P^T C^-1 P)^-1 P^T C^-1 y
with variance
(dx)^2 = (P^T C^-1 P)^-1

The ratio of x / dx is the significance is x / (dx) = P^T C^-1 y / sqrt(P^T C^-1 P) .  The term in the denominator with P^T P is the sum of the square of the PSF.

Similarly, I claim if you compute signal(i) = sum_j PSF(i, j) * image(j) * weight(j) and ask what the variance is, you likewise get a PSF^2 term that would correspond to convolving with the square of the kernel.